merino/providers/top_picks/provider.py

```
@@ -0,0 +1,97 @@
+import logging
+from collections import defaultdict
+
+from merino.providers.base import BaseProvider, BaseSuggestion, SuggestionRequest
+from merino.providers.top_picks.backends.protocol import TopPicksBackend, TopPicksData
+
+logger = logging.getLogger(__name__)
+
+
+class Suggestion(BaseSuggestion):
+    """Model for Top Pick Suggestion."""
+
+    block_id: int
+    is_top_pick: bool
+
+
+class Provider(BaseProvider):
+    """Top Pick Suggestion Provider."""
+
+    top_picks_data: TopPicksData
+    primary_index: defaultdict = defaultdict(list)
+    secondary_index: defaultdict = defaultdict(list)
+    short_domain_index: defaultdict = defaultdict(list)
+    results: list[Suggestion]
+    query_min: int
+    query_max: int
+
+    def __init__(
+        self,
+        backend: TopPicksBackend,
+        name: str,
+        enabled_by_default: bool = False,
+    ) -> None:
+        self.backend = backend
+        self._name = name
+        self._enabled_by_default = enabled_by_default
+
+    async def initialize(self) -> None:
+        """Initialize the provider."""
+        try:
+            # Fetch Top Picks suggestions from domain list.
+            self.top_picks_data: TopPicksData = await self.backend.fetch()
+            self.primary_index = self.top_picks_data.primary_index
+            self.secondary_index = self.top_picks_data.secondary_index
+            self.short_domain_index = self.top_picks_data.short_domain_index
+            # Convert result dictionaries into Suggestion objects.
+            self.results: list[Suggestion] = [
+                Suggestion(**result) for result in self.top_picks_data.results
+            ]
+            self.query_min = self.top_picks_data.index_char_range[0]
+            self.query_max = self.top_picks_data.index_char_range[1]
+            self.query_char_limit = self.top_picks_data.query_char_limit
+            self.firefox_char_limit = self.top_picks_data.firefox_char_limit
+
+        except Exception as e:
```

Since `backend.fetch()` will only raise `TopPicksError` errors, let's capture those here instead. Also, apart from logging, we might also want to re-throw the exception to let provider initializer handle it rather than silencing it.

---

Good call. Actually, I was wondering about the motivation of `TopPicksError`, it doesn't seem to have any specific exceptional behaviors for this backend. Should we just use `BackendError` instead?

---

You could also do that. I think the name/type of the error brings clarity 'TopPicksError', you know exactly which backend raised that, but it's not essential.

---

Ok, I will catch the `BackendError here then raise the Error. Will update.